man/boobmsg.1
.\" -*- coding: utf-8 -*-
.\" This file was generated automatically by tools/make_man.sh.
.TH BOOBMSG 1 "13 February 2020" "boobmsg 2\&.0"
.SH NAME
boobmsg \- send and receive message threads
.SH SYNOPSIS
.B boobmsg
[\-dqv] [\-b \fIbackends\fR] [\-cnfs] [\fIcommand\fR [\fIarguments\fR..]]
.br
.B boobmsg
[\-\-help] [\-\-version]

.SH DESCRIPTION
.LP

Console application allowing to send messages on various websites and to display message threads and contents.

.SS Supported websites:
* aum ("Adopte un Mec" French dating website)
.br
* blogspot (Blog reader for blogspot. Read\-only and without comments.)
.br
* bnporc (BNP Paribas)
.br
* dlfp (Da Linux French Page news website)
.br
* feedly (handle the popular RSS reading service Feedly)
.br
* freemobile (Free Mobile website)
.br
* guerrillamail (GuerrillaMail temp mailbox)
.br
* happn (Happn dating mobile application)
.br
* hds (Histoires de Sexe French erotic novels)
.br
* inrocks (Les Inrocks French news website)
.br
* lefigaro (Le Figaro French newspaper website)
.br
* liberation (Libération newspaper website)
.br
* mailinator (mailinator temp mailbox)
.br
* minutes20 (20 Minutes French newspaper website)
.br
* newsfeed (Loads RSS and Atom feeds from any website)
.br
* okc (OkCupid)
.br
* openedx (Discussions on OpenEDX\-powered coursewares)
.br
* phpbb (phpBB forum)
.br
* playme (PlayMe dating mobile application)
.br
* presseurop (Presseurop website)
.br
* reddit (reddit website)
.br
* tapatalk (Tapatalk\-compatible sites)
.br
* taz (Taz newspaper website)
.br
* tinder (Tinder dating mobile application)
.br
* twitter (twitter website)
.SH BOOBMSG COMMANDS
.TP
\fBdebug\fR
.br
Launch a debug Python shell
.TP
\fBexport_all\fR
.br
Export All threads
.TP
\fBexport_thread\fR \fIID\fR
.br
Export the thread identified by ID
.TP
\fBlist\fR
.br
Display all threads.
.br

.br
Default is limited to 10 results.
.TP
\fBphotos\fR \fIID\fR
.br
Display photos of a profile
.TP
\fBpost\fR \fIRECEIVER\fR@\fIBACKEND\fR[,\fIRECEIVER\fR@\fIBACKEND\fR[...]] [\fITEXT\fR]
.br
Post a message to the specified receivers.
.br
Multiple receivers are separated by a comma.
.br

.br
If no text is supplied on command line, the content of message is read on stdin.
.TP
\fBprofile\fR \fIID\fR
.br
Display a profile
.TP
\fBshow\fR \fIMESSAGE\fR
.br
Read a message
.TP
\fBstatus\fR
.br
Display status information about a backend.
.SH WEBOOB COMMANDS
.TP
\fBbackends\fR [\fIACTION\fR] [\fIBACKEND_NAME\fR]...
.br
Select used backends.
.br

.br
ACTION is one of the following (default: list):
.br
* enable         enable given backends
.br
* disable        disable given backends
.br
* only           enable given backends and disable the others
.br
* list           list backends
.br
* add            add a backend
.br
* register       register a new account on a website
.br
* edit           edit a backend
.br
* remove         remove a backend
.br
* list\-modules   list modules
.TP
\fBcd\fR [\fIPATH\fR]
.br
Follow a path.
.br
".." is a special case and goes up one directory.
.br
"" is a special case and goes home.
.TP
\fBcondition\fR [\fIEXPRESSION\fR | off]
.br
If an argument is given, set the condition expression used to filter the results. See CONDITION section for more details and the expression.
.br
If the "off" value is given, conditional filtering is disabled.
.br

.br
If no argument is given, print the current condition expression.
.TP
\fBcount\fR [\fINUMBER\fR | off]
.br
If an argument is given, set the maximum number of results fetched.
.br
NUMBER must be at least 1.
.br
"off" value disables counting, and allows infinite searches.
.br

.br
If no argument is given, print the current count value.
.TP
\fBformatter\fR [list | \fIFORMATTER\fR [\fICOMMAND\fR] | option \fIOPTION_NAME\fR [on | off]]
.br
If a FORMATTER is given, set the formatter to use.
.br
You can add a COMMAND to apply the formatter change only to
.br
a given command.
.br

.br
If the argument is "list", print the available formatters.
.br

.br
If the argument is "option", set the formatter options.
.br
Valid options are: header, keys.
.br
If on/off value is given, set the value of the option.
.br
If not, print the current value for the option.
.br

.br
If no argument is given, print the current formatter.
.TP
\fBlogging\fR [\fILEVEL\fR]
.br
Set logging level.
.br

.br
Availables: debug, info, warning, error.
.br
* quiet is an alias for error
.br
* default is an alias for warning
.TP
\fBls\fR [\-d] [\-\fIU\fR] [\fIPATH\fR]
.br
List objects in current path.
.br
If an argument is given, list the specified path.
.br
Use \-U option to not sort results. It allows you to use a "fast path" to
.br
return results as soon as possible.
.br
Use \-d option to display information about a collection (and to not
.br
display the content of it). It has the same behavior than the well
.br
known UNIX "ls" command.
.br

.br
Default is limited to 40 results.
.TP
\fBquit\fR
.br
Quit the application.
.TP
\fBselect\fR [\fIFIELD_NAME\fR]... | "$direct" | "$full"
.br
If an argument is given, set the selected fields.
.br
$direct selects all fields loaded in one http request.
.br
$full selects all fields using as much http requests as necessary.
.br

.br
If no argument is given, print the currently selected fields.

.SH OPTIONS
.TP
\fB\-\-version\fR
show program's version number and exit
.TP
\fB\-h\fR, \fB\-\-help\fR
show this help message and exit
.TP
\fB\-b BACKENDS\fR, \fB\-\-backends=BACKENDS\fR
what backend(s) to enable (comma separated)
.TP
\fB\-e EXCLUDE_BACKENDS\fR, \fB\-\-exclude\-backends=EXCLUDE_BACKENDS\fR
what backend(s) to exclude (comma separated)
.TP
\fB\-I\fR, \fB\-\-insecure\fR
do not validate SSL
.TP
\fB\-\-nss\fR
Use NSS instead of OpenSSL
.TP
\fB\-\-auto\-update\fR
Automatically check for updates when a bug in a module is encountered

.SH BOOBMSG OPTIONS
.TP
\fB\-E\fR, \fB\-\-accept\-empty\fR
Send messages with an empty body.
.TP
\fB\-t TITLE\fR, \fB\-\-title=TITLE\fR
For the "post" command, set a title to message

.SH LOGGING OPTIONS
.TP
\fB\-d\fR, \fB\-\-debug\fR
display debug messages. Set up it twice to more verbosity
.TP
\fB\-q\fR, \fB\-\-quiet\fR
display only error messages
.TP
\fB\-v\fR, \fB\-\-verbose\fR
display info messages
.TP
\fB\-\-logging\-file=LOGGING_FILE\fR
file to save logs
.TP
\fB\-a\fR, \fB\-\-save\-responses\fR
save every response
.TP
\fB\-\-export\-session\fR
log browser session cookies after login

.SH RESULTS OPTIONS
.TP
\fB\-c CONDITION\fR, \fB\-\-condition=CONDITION\fR
filter result items to display given a boolean expression. See CONDITION section
for the syntax
.TP
\fB\-n COUNT\fR, \fB\-\-count=COUNT\fR
limit number of results (from each backends)
.TP
\fB\-s SELECT\fR, \fB\-\-select=SELECT\fR
select result item keys to display (comma separated)

.SH FORMATTING OPTIONS
.TP
\fB\-f FORMATTER\fR, \fB\-\-formatter=FORMATTER\fR
select output formatter (atom, csv, htmltable, json, json_line, msg, msglist,
multiline, profile, simple, table, webkit, xhtml)
.TP
\fB\-\-no\-header\fR
do not display header
.TP
\fB\-\-no\-keys\fR
do not display item keys
.TP
\fB\-O OUTFILE\fR, \fB\-\-outfile=OUTFILE\fR
file to export result

.SH CONDITION
The \-c and \-\-condition is a flexible way to filter and get only interesting results. It supports conditions on numerical values, dates, durations and strings. Dates are given in YYYY\-MM\-DD or YYYY\-MM\-DD HH:MM format. Durations look like XhYmZs where X, Y and Z are integers. Any of them may be omitted. For instance, YmZs, XhZs or Ym are accepted.
The syntax of one expression is "\fBfield operator value\fR". The field to test is always the left member of the expression.
.LP
The field is a member of the objects returned by the command. For example, a bank account has "balance", "coming" or "label" fields.
.SS The following operators are supported:
.TP
=
Test if object.field is equal to the value.
.TP
!=
Test if object.field is not equal to the value.
.TP
>
Test if object.field is greater than the value. If object.field is date, return true if value is before that object.field.
.TP
<
Test if object.field is less than the value. If object.field is date, return true if value is after that object.field.
.TP
|
This operator is available only for string fields. It works like the Unix standard \fBgrep\fR command, and returns True if the pattern specified in the value is in object.field.
.SS Expression combination
.LP
You can make a expression combinations with the keywords \fB" AND "\fR, \fB" OR "\fR an \fB" LIMIT "\fR.
.LP
The \fBLIMIT\fR keyword can be used to limit the number of items upon which running the expression. \fBLIMIT\fR can only be placed at the end of the expression followed by the number of elements you want.
.SS Examples:
.nf
.B boobank ls \-\-condition 'label=Livret A'
.fi
Display only the "Livret A" account.
.PP
.nf
.B boobank ls \-\-condition 'balance>10000'
.fi
Display accounts with a lot of money.
.PP
.nf
.B boobank history account@backend \-\-condition 'label|rewe'
.fi
Get transactions containing "rewe".
.PP
.nf
.B boobank history account@backend \-\-condition 'date>2013\-12\-01 AND date<2013\-12\-09'
.fi
Get transactions betweens the 2th December and 8th December 2013.
.PP
.nf
.B boobank history account@backend \-\-condition 'date>2013\-12\-01  LIMIT 10'
.fi
Get transactions after the 2th December in the last 10 transactions

.SH COPYRIGHT
Copyright(C) 2010-2020 Christophe Benz
.LP
For full copyright information see the COPYING file in the weboob package.
.LP
.RE
.SH FILES
"~/.config/weboob/backends" 

.SH SEE ALSO
Home page: http://weboob.org/applications/boobmsg


ChangeLog
Weboob 2.0 (2020-02-13)

	Applications: boobank
	* keep Value for TransferStep and AddRecipientStep
	* declare utf-8 in OFX formatter
	* give account object to iter_transfer_recipient

	Applications: boobill
	* using a persistent storage for browser state

	Applications: boobsize
	* fix name

	Applications: translaboob
	* use babel to detect language names and pass ISO-639 codes

	Applications: weboobdebug
	* weboob-debug: fix name

	Applications: weboobmain
	* move scripts/weboob to a dedicated dir in weboob/applications

	Applications: weboobrepos
	* weboob-repos: port to python3

	Browser
	* set a parent logger to AbstractElement
	* Allows SeleniumBrowser to use MANUAL proxies
	* AUTHENTICATION_METHODS set in init
	* delete prefix 'handle_' on authentication methods
	* clean authentication config keys after login
	* make variable name more understandable
	* make get_expire stronger
	* TWOFA_DURATION for the state expire date
	* better names
	* set expire from logged_date
	* make authentication methods more flexible
	* handle both 2FA and regular login methods
	* add possibility to clear specific cookies
	* handle NeedInteractiveFor2FA
	* add documentation about otp, sms and resume
	* add TwoFactorBrowser to handle multiple 2FA
	* Set default logging level to ERROR for selenium
	* Added the ability to overrride *Elements' env copy
	* Allow passing custom headers in Form.submit()
	* set attribute 'weboob' on browser
	* Normalize encodings
	* BrowserTooManyRequests: Remove class attributes for date helpers
	* BrowserTooManyRequests: Raise a TypeError if the next_try is not a datetime
	* BrowserTooManyRequests: Check whether the next_try is None after the date check
	* BrowserTooManyRequests: Add message parameter
	* [weboob/browser/browser+pages] Fixed doctest BASEURL
	* add some helpers to TooManyRequests
	* add rate limiting exception
	* add build_refresh_token_parameters method
	* python3 compatibility
	* send redirect_uri when using the refresh_token
	* allow multiple inheritance with AbstractBrowser
	* Implemented wait_xpath_invisible
	* Avoid crash when no BASEURL is set
	* add klass attribute to decorated @method
	* give state to the module at initialization
	* introduce BrowserParamURL for passing params implictly
	* Implement the empty_xpath attribute
	* accept new parameter proxy_headers (but don't use it)
	* Browser takes a new parameter 'proxy_headers'
	* if use of refresh_token fails, forget it
	* support when token_type key is missing
	* if access_token_expire is None, the access_token has no expiration
	* in case we got a 401 response, forget the access_token
	* fix compatibility with python3
	* Handle generic error for bad url callback
	* Repair refresh token
	* StatesMixin cookies should be str
	* OAuth2PKCEMixin codes shouldn't be bytes
	* Rewrite/improve PagesBrowser example

	Browser: Filters
	* Add the format .Italian() to CleanDecimal
	* New keyword argument to remove accented characters on CleanText filter.
	* Changed Capitalize filter to Title
	* Updated & Reordered __all__
	* Add dates to FormValue
	* Add color to FormValue
	* Add number/range type to FormValue, and tests
	* MultiJoin accepts types other than str
	* Regexp can use regex library
	* Allow all dateutil.parser arguments in DateTime
	* Strict dates by default
	* Catch incomplete dates
	* Explicit error message
	* Add MultiJoin
	* Join can use default argument
	* Add doctests to Join
	* TableCell(colspan=True) is now the only behavior desired
	* add FromTimestamp to convert to datetime
	* Fix error message for FormValue
	* Add new types for FormValue
	* Add MapIn filter, that allow the pattern of a selected value to another value using a dict.
	* Reorganize imports / add missing to __all__
	* Add Coalesce

	Capabilities
	* rename accepted_doc_types attribute of CapDocument
	* Add KIID to DocumentTypes
	* Added new capability CapBands
	* Keep only PER
	* add PER account type
	* HousingPhoto inherits BaseImage
	* use PostalAddress and GeoCoords types where possible
	* introduce weboob.capabilities.address for a few object types
	* Add firstname and lastname field
	* add capability_to_string function to get capability name
	* Add new SAR currency
	* fix missing TransferInvalidOTP export
	* add TransferInvalidOTP exception
	* use StringField standard default value for transfer beneficiary label and number
	* Indentation cosmetic changes
	* Add new Investment attributes
	* Add new document type 'income_tax'
	* Add field 'original_gross_amount' to Transaction
	* Add field 'gross_amount' to Transaction
	* Remove performance_history default value
	* Added new attributes for employee savings accounts
	* Add RecaptchaV3Job to captcha capability
	* add support for commission in currency in bank/Transaction
	* add new add recipient exception for invalid otp
	* add has_file field to Document class
	* Add the relationship between the credentials owner (PSU) and the account
	* Add bdate. Represent the date that the transaction appear in the website.
	* prevent casting id from None to 'None' in constructor

	Documentation
	* add pip install
	* fix compatibility with sphinx 2.0
	* do not use deprecated defindex.html anymore
	* Use correct word

	General
	* New airparif module (CapGauge)
	* New aviva module (CapBank, CapBankWealth)
	* New cesu module (CapDocument)
	* New degiro module (CapBank, CapBankWealth)
	* New hsbchk module (CapBank)
	* New lendosphere module (CapBank, CapBankWealth)
	* New lucca module (CapCalendarEvent, CapDocument)
	* New lunchr module (CapBank)
	* New metalarchives module (CapBands)
	* New pajemploi module (CapDocument)
	* New ticketscesu module (CapBank)
	* New wiseed module (CapBank, CapBankWealth)
	* Deleted biplan module
	* Deleted figgo module
	* Deleted lolix module

weboob/applications/boobmsg/boobmsg.py
# -*- coding: utf-8 -*-

# Copyright(C) 2010-2011  Christophe Benz
#
# This file is part of weboob.
#
# weboob is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# weboob is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with weboob. If not, see <http://www.gnu.org/licenses/>.

from __future__ import print_function

import os
import datetime
import hashlib

from tempfile import NamedTemporaryFile
from lxml import etree

from weboob.core import CallErrors
from weboob.capabilities.base import empty
from weboob.capabilities.messages import CapMessages, Message, Thread
from weboob.capabilities.account import CapAccount
from weboob.capabilities.contact import CapContact
from weboob.tools.application.repl import ReplApplication, defaultcount
from weboob.tools.application.formatters.iformatter import IFormatter
from weboob.tools.html import html2text


__all__ = ['Boobmsg']


class AtomFormatter(IFormatter):
    MANDATORY_FIELDS = ('title', 'date', 'sender', 'content')

    def start_format(self, **kwargs):
        self.output(u'<?xml version="1.0" encoding="utf-8"?><feed xmlns="http://www.w3.org/2005/Atom"')
        self.output(u'xmlns:dc="http://purl.org/dc/elements/1.1/">\n')
        self.output(u'<title type="text">Atom feed by Weboob</title>')  # TODO : get backend name
        self.output(u'<updated>%s</updated>' % datetime.datetime.now().strftime("%Y-%m-%dT%H:%M:%SZ"))
        m = hashlib.md5()
        m.update(datetime.datetime.now().strftime("%Y-%m-%dT%H:%M:%SZ"))
        self.output(u'<id>urn:md5:%s</id>' % m.hexdigest())

    def format_obj(self, obj, alias):
        elem = etree.Element('entry')

        title = etree.Element('title')
        title.text = obj.title
        elem.append(title)

        id = etree.Element('id')
        m = hashlib.md5()
        m.update(obj.content.encode('utf8', 'ascii'))
        id.text = "urn:md5:%s" % m.hexdigest()
        elem.append(id)

        link = etree.Element('link')
        link.attrib["href"] = obj.thread.id
        link.attrib["title"] = obj.title
        link.attrib["type"] = "text/html"
        elem.append(link)

        author = etree.Element('author')
        name = etree.Element('name')
        if obj.sender:
            name.text = obj.sender
        else:
            name.text = obj.backend
        author.append(name)
        elem.append(author)

        date = etree.Element('updated')
        date.text = obj.date.strftime("%Y-%m-%dT%H:%M:%SZ")
        elem.append(date)

        content = etree.Element('content')
        content.text = obj.content
        content.attrib["type"] = "html"
        elem.append(content)

        return etree.tostring(elem, pretty_print=True)

    def flush(self):
        self.output(u'</feed>')


class XHtmlFormatter(IFormatter):
    MANDATORY_FIELDS = ('title', 'date', 'sender', 'signature', 'content')

    def format_obj(self, obj, alias):
        result  = "<div>\n"
        result += "<h1>%s</h1>" % (obj.title)
        result += "<dl>"
        result += "<dt>Date</dt><dd>%s</dd>" % (obj.date.strftime('%Y-%m-%d %H:%M'))
        result += "<dt>Sender</dt><dd>%s</dd>" % (obj.sender)
        result += "<dt>Signature</dt><dd>%s</dd>" % (obj.signature)
        result += "</dl>"
        result += "<div>%s</div>" % (obj.content)
        result += "</div>\n"
        return result


class MessageFormatter(IFormatter):
    MANDATORY_FIELDS = ('title', 'date', 'sender', 'signature', 'content')

    def format_obj(self, obj, alias):
        result = u'%sTitle:%s %s\n' % (self.BOLD,
                                       self.NC, obj.title)
        result += u'%sDate:%s %s\n' % (self.BOLD,
                                       self.NC, obj.date.strftime('%Y-%m-%d %H:%M'))
        result += u'%sFrom:%s %s\n' % (self.BOLD,
                                       self.NC, obj.sender)
        if hasattr(obj, 'receivers') and obj.receivers:
            result += u'%sTo:%s %s\n' % (self.BOLD,
                                         self.NC,
                                         ', '.join(obj.receivers))

        if obj.flags & Message.IS_HTML:
            content = html2text(obj.content)
        else:
            content = obj.content

        result += '\n%s' % content

        if obj.signature:
            if obj.flags & Message.IS_HTML:
                signature = html2text(obj.signature)
            else:
                signature = obj.signature

            result += '\n-- \n%s' % signature
        return result


class MessagesListFormatter(IFormatter):
    MANDATORY_FIELDS = ()
    count = 0
    _list_messages = False

    def flush(self):
        self.count = 0

    def format_obj(self, obj, alias):
        if not self._list_messages:
            return self.format_dict_thread(obj, alias)
        else:
            return self.format_dict_messages(obj, alias)

    def format_dict_thread(self, obj, alias):
        self.count += 1
        if self.interactive:
            result = u'%s* (%d) %s (%s)%s' % (self.BOLD,
                                              self.count,
                                              obj.title, obj.backend,
                                              self.NC)
        else:
            result = u'%s* (%s) %s%s' % (self.BOLD, obj.id,
                                         obj.title,
                                         self.NC)
        if obj.date:
            result += u'\n             %s' % obj.date
        return result

    def format_dict_messages(self, obj, alias):
        if obj.flags == Thread.IS_THREADS:
            depth = 0
        else:
            depth = -1

        result = self.format_message(obj.backend, obj.root, depth)
        return result

    def format_message(self, backend, message, depth=0):
        if not message:
            return u''
        self.count += 1

        flags = '['
        if message.flags & message.IS_UNREAD:
            flags += 'N'
        else:
            flags += '-'
        if message.flags & message.IS_NOT_RECEIVED:
            flags += 'U'
        elif message.flags & message.IS_RECEIVED:
            flags += 'R'
        else:
            flags += '-'
        flags += ']'

        if self.interactive:
            result = u'%s%s* (%d)%s %s <%s> %s (%s)\n' % (depth * '  ',
                                                          self.BOLD,
                                                          self.count,
                                                          self.NC,
                                                          flags,
                                                          message.sender,
                                                          message.title,
                                                          backend)
        else:
            result = u'%s%s* (%s.%s@%s)%s %s <%s> %s\n' % (depth * '  ',
                                                           self.BOLD,
                                                           message.thread.id,
                                                           message.id,
                                                           backend,
                                                           self.NC,
                                                           flags,
                                                           message.sender,
                                                           message.title)
        if message.children:
            if depth >= 0:
                depth += 1
            for m in message.children:
                result += self.format_message(backend, m, depth)
        return result


class ProfileFormatter(IFormatter):
    def flush(self):
        pass

    def format_obj(self, obj, alias=None):
        return obj.get_text()


class Boobmsg(ReplApplication):
    APPNAME = 'boobmsg'
    VERSION = '2.1'
    COPYRIGHT = 'Copyright(C) 2010-YEAR Christophe Benz'
    DESCRIPTION = "Console application allowing to send messages on various websites and " \
                  "to display message threads and contents."
    SHORT_DESCRIPTION = "send and receive message threads"
    CAPS = CapMessages
    EXTRA_FORMATTERS = {'msglist':  MessagesListFormatter,
                        'msg':      MessageFormatter,
                        'xhtml':    XHtmlFormatter,
                        'atom':     AtomFormatter,
                        'profile' : ProfileFormatter,
                       }
    COMMANDS_FORMATTERS = {'list':          'msglist',
                           'show':          'msg',
                           'export_thread': 'msg',
                           'export_all':    'msg',
                           'ls':            'msglist',
                           'profile':       'profile',
                          }

    def add_application_options(self, group):
        group.add_option('-E', '--accept-empty',  action='store_true',
                         help='Send messages with an empty body.')
        group.add_option('-t', '--title', action='store',
                         help='For the "post" command, set a title to message',
                         type='string', dest='title')

    def load_default_backends(self):
        self.load_backends(CapMessages, storage=self.create_storage())

    def main(self, argv):
        self.load_config()
        return ReplApplication.main(self, argv)

    def do_status(self, line):
        """
        status

        Display status information about a backend.
        """
        if len(line) > 0:
            backend_name = line
        else:
            backend_name = None

        results = {}
        for field in self.do('get_account_status',
                             backends=backend_name,
                             caps=CapAccount):
            if field.backend in results:
                results[field.backend].append(field)
            else:
                results[field.backend] = [field]

        for name, fields in results.items():
            print(':: %s ::' % name)
            for f in fields:
                if f.flags & f.FIELD_HTML:
                    value = html2text(f.value)
                else:
                    value = f.value
                print('%s: %s' % (f.label, value))
            print('')

    def do_post(self, line):
        """
        post RECEIVER@BACKEND[,RECEIVER@BACKEND[...]] [TEXT]

        Post a message to the specified receivers.
        Multiple receivers are separated by a comma.

        If no text is supplied on command line, the content of message is read on stdin.
        """
        receivers, text = self.parse_command_args(line, 2, 1)
        if text is None:
            text = self.acquire_input()

        if not self.options.accept_empty and not text.strip():
            self.logger.warning(u'The message body is empty, use option --accept_empty to send empty messages')
            return

        for receiver in receivers.strip().split(','):
            receiver, backend_name = self.parse_id(receiver.strip(),
                                                   unique_backend=True)
            if not backend_name and len(self.enabled_backends) > 1:
                self.logger.warning(u'No backend specified for receiver "%s": message will be sent with all the '
                    'enabled backends (%s)' % (receiver,
                    ','.join(backend.name for backend in self.enabled_backends)))

            if '.' in receiver:
                # It's a reply
                thread_id, parent_id = receiver.rsplit('.', 1)
            else:
                # It's an original message
                thread_id = receiver
                parent_id = None
                try:
                    thread_id = self.threads[int(thread_id) - 1].id
                except (IndexError,ValueError):
                    pass

            thread = Thread(thread_id)
            message = Message(thread,
                              0,
                              title=self.options.title,
                              parent=Message(thread, parent_id) if parent_id else None,
                              content=text)

            try:
                self.do('post_message', message, backends=backend_name).wait()
            except CallErrors as errors:
                self.bcall_errors_handler(errors)
            else:
                if self.interactive:
                    print('Message sent sucessfully to %s' % receiver)

    threads = []
    messages = []

    @defaultcount(10)
    def do_list(self, arg):
        """
        list

        Display all threads.
        """
        if len(arg) > 0:
            try:
                thread = self.threads[int(arg) - 1]
            except (IndexError, ValueError):
                id, backend_name = self.parse_id(arg)
            else:
                id = thread.id
                backend_name = thread.backend

            self.messages = []
            cmd = self.do('get_thread', id, backends=backend_name)
            self.formatter._list_messages = True
        else:
            self.threads = []
            cmd = self.do('iter_threads')
            self.formatter._list_messages = False

        self.start_format()
        for thread in cmd:
            if not thread:
                continue
            if len(arg) > 0:
                for m in thread.iter_all_messages():
                    if not m.backend:
                        m.backend = thread.backend
                    self.messages.append(m)
            else:
                self.threads.append(thread)
            self.format(thread)

    def do_export_all(self, arg):
        """
        export_all

        Export All threads
        """

        def func(backend):
            for thread in backend.iter_threads():
                if not thread:
                    continue
                t = backend.fillobj(thread, None)
                for msg in t.iter_all_messages():
                    yield msg

        self.start_format()
        for msg in self.do(func):
            self.format(msg)

    def do_export_thread(self, arg):
        """
        export_thread ID

        Export the thread identified by ID
        """
        try:
            thread = self.threads[int(arg) - 1]
        except (IndexError, ValueError):
            _id, backend_name = self.parse_id(arg)
        else:
            _id = thread.id
            backend_name = thread.backend

        cmd = self.do('get_thread', _id, backends=backend_name)
        self.start_format()
        for thread in cmd:
            if thread is not None:
                thread, = self.do('fillobj', thread, None, backends=thread.backend)
                for msg in thread.iter_all_messages():
                    msg, = self.do('fillobj', msg, None, backends=thread.backend)
                    self.format(msg)

    def do_show(self, arg):
        """
        show MESSAGE

        Read a message
        """
        message = None
        if len(arg) == 0:
            print('Please give a message ID.', file=self.stderr)
            return 2

        try:
            message = self.messages[int(arg) - 1]
        except (IndexError, ValueError):
            # The message is not is the cache, we have now two cases:
            # 1) the user uses a number to get a thread in the cache
            # 2) the user gives a thread id
            try:
                thread = self.threads[int(arg) - 1]
                if not empty(thread.root):
                    message = thread.root
                else:
                    for thread in self.do('get_thread', thread.id, backends=thread.backend):
                        if thread is not None:
                            if not thread.root:
                                thread, = self.do('fillobj', thread, ('root',), backends=thread.backend)
                            message = thread.root
            except (IndexError, ValueError):
                _id, backend_name = self.parse_id(arg)
                for thread in self.do('get_thread', _id, backends=backend_name):
                    if thread is not None:
                        if not thread.root:
                            thread, = self.do('fillobj', thread, ('root',), backends=thread.backend)
                        message = thread.root
        if not empty(message):
            self.start_format()
            self.format(message)
            self.weboob.do('set_message_read', message, backends=message.backend)
            return
        else:
            print('Message not found', file=self.stderr)
            return 3

    def do_profile(self, id):
        """
        profile ID

        Display a profile
        """
        _id, backend_name = self.parse_id(id, unique_backend=True)

        found = 0
        for contact in self.do('get_contact', _id, backends=backend_name, caps=CapContact):
            if contact:
                self.format(contact)
                found = 1

        if not found:
            self.logger.error(u'Profile not found')

    def do_photos(self, id):
        """
        photos ID

        Display photos of a profile
        """
        photo_cmd = self.config.get('photo_viewer')
        if photo_cmd is None:
            print("Configuration error: photo_viewer is undefined", file=self.stderr)
            return

        _id, backend_name = self.parse_id(id, unique_backend=True)

        found = 0
        for contact in self.do('get_contact', _id, backends=backend_name):
            if contact:
                # Write photo to temporary files
                tmp_files = []
                for photo in contact.photos.values():
                    suffix = '.jpg'
                    if '.' in photo.url.split('/')[-1]:
                        suffix = '.%s' % photo.url.split('/')[-1].split('.')[-1]
                    f = NamedTemporaryFile(suffix=suffix)

                    photo = self.weboob[contact.backend].fillobj(photo, 'data')
                    f.write(photo.data)
                    tmp_files.append(f)
                os.system(photo_cmd % ' '.join([file.name for file in tmp_files]))
                found = 1

        if not found:
            self.logger.error(u'Profile not found')
